cf_xarray/parametric.py

```diff
@@ -0,0 +1,691 @@
+            raise ValueError(
+                f"The standard name for the {x!r} variable is not available."
+            ) from None
+
+    for x, y in ocean_stdname_map.items():
+        check_term = "".join(
+            [
+                y[i]
+                for i, j in sorted(kwargs.items(), key=lambda x: x[0])
+                if j is not None
+            ]
+        )
+
+        if search_term == check_term:
+            found_stdname = x
+
+            break
+
+    if found_stdname is None:
+        stdnames = ", ".join(
+            [y["standard_name"] for _, y in sorted(kwargs.items(), key=lambda x: x[0])]
+        )
+
+        raise ValueError(
+            f"Could not derive standard name from combination of {stdnames}."
+        )
+
+    return found_stdname
+
+
+def check_requirements(func, terms):
+    """Checks terms against function requirements.
+
+    Uses `func` argument specification as requirements and checks terms against this.
+    Postitional arguments without a default are required but when a default value is
+    provided the arguement is considered optional. Atleast one optional argument must
+    be present (special case for atmosphere_hybrid_sigma_pressure_coordinate).
+
+    Parameters
+    ----------
+    func : function
+        Function to check requirements.
+    terms : list
+        List of terms to check `func` requirements against.
+
+    Raises
+    ------
+    KeyError
+        If `terms` is empty or missing required/optional terms.
+    """
+    if not isinstance(terms, set):
+        terms = set(terms)
+
+    spec = inspect.getfullargspec(func)
+
+    args = spec.args or []
+
+    if len(terms) == 0:
+        raise KeyError(f"Required terms {', '.join(sorted(args))} absent in dataset.")
+
+    # handle case insensitive
+    terms = {x.lower() for x in terms}
+
+    n = len(spec.defaults or [])
+
+    # last `n` arguments are optional
+    opt = set(args[len(args) - n :])
+
+    req = set(args) - opt
+
+    # req must all be present in terms
+    if (req & terms) != req:
+        req_diff = sorted(req - terms)
+
+        raise KeyError(
+            f"Required terms {', '.join(req_diff)} are absent in the dataset."
+        )
+
+    # if there are optional arguments check that atleast one
+    # is in the intersection, only required for
+    # atmosphere_hybrid_sigma_pressure_coordinate
+    if len(opt) > 0 and len(opt & terms) == 0:
+        raise KeyError(
+            f"Atleast one optional term {', '.join(sorted(opt))} is absent in the dataset."
+        )
+
+
+def func_from_stdname(stdname):
+    """Get function from module.
+
+    Uses `stdname` to return function from module.
+
+    Parameters
+    ----------
+    stdname : str
+        Name of the function.
+
+    Raises
+    ------
+    AttributeError
+        If a function name `stdname` is not in the module.
+    """
+    m = sys.modules[__name__]
+
+    return getattr(m, stdname)
+
+
+def derive_dimension_order(output_order, **dim_map):
```

Comment:
This is quite clever, but I'd prefer if the user just did the transpose on their own after this decoding.

Reply:
Yes with Xarray we care much less about dimension order.

Reply:
@jasonb5 / @dcherian – thanks for all of your work on this. I wanted to weigh-in as a user. It would be unexpected to get a matrix of `[time, lat, lon, *plev*]` instead of `[time, *plev*, lat, lon]`. I think users would prefer the former (matches most climate data, CF conventions, and almost certainly the order of the original data, which is typically `[time, *lev*, lat lon]`. 

If this doesn't match cf-xarray's style, xcdat could apply this functionality after calling cf-xarray's transform.

Reply:
Yes xcdat could do `.cf.transpose('T', 'Z', 'Y', 'X')`

> almost certainly the order of the original data, which is typically [time, *lev*, lat lon]).

Is this ordering not preserved here?

Reply:
I just did a quick check: it looks like it goes from (`ds.ta.dims`):

`('time', 'lev', 'lat', 'lon')`

to (`p = ds.a * ds.p0 + ds.b * ds.ps; p.dims`):

`('lev', 'time', 'lat', 'lon')`

In other words, the dataarray of interest is `[time, lev, lat, lon]`, but the transform will put `lev` at the front `[lev, time, lat, lon]`